<p>In this sub-section, I will be identifying and removing duplicated data within my merged dataframe. Duplicated can have a negative effect on the Data Analytics which I will be performing. Therefore, it is of the utmost importance that duplicated data be identified and addressed.</p>

<hr/>
<h3>2.4.1 Removing Duplicated Data Column - Dataset 1</h3>

<p>As previously observed, there are Duplicated Data Columns which are listed below.</p>
<br/>
<ul>
    <li>gross_monthly_mean.1</li>
    <li>gross_monthly_median.1</li>
    <li>gross_mthly_25_percentile.1</li>
    <li>gross_mthly_75_percentile.1</li>
</ul>
<br/>
<p>With the Duplicated Data Column listed, we will proceed to manually drop the Data Column in the cell below.</p>

In [126]:
# ---------- Dropping Duplicated Data Columns ---------- #
df1.drop(["gross_monthly_mean.1", "gross_monthly_median.1", "gross_mthly_25_percentile.1", "gross_mthly_75_percentile.1"], axis=1, inplace=True)

# ---------- Displaying DataFrame ---------- #
df1.head(5).style.background_gradient(cmap="Blues")

,year,university,school,degree,employment_rate_overall,employment_rate_ft_perm,basic_monthly_mean,basic_monthly_median,gross_monthly_mean,gross_monthly_median,gross_mthly_25_percentile,gross_mthly_75_percentile
0,2013,Nanyang Technological University,College of Business (Nanyang Business School),Accountancy and Business,97.4,96.1,3701,3200,3727,3350,2900,4000
1,2013,Nanyang Technological University,College of Business (Nanyang Business School),Accountancy (3-yr direct Honours Programme),97.1,95.7,2850,2700,2938,2700,2700,2900
2,2013,Nanyang Technological University,College of Business (Nanyang Business School),Business (3-yr direct Honours Programme),90.9,85.7,3053,3000,3214,3000,2700,3500
3,2013,Nanyang Technological University,College of Business (Nanyang Business School),Business and Computing,87.5,87.5,3557,3400,3615,3400,3000,4100
4,2013,Nanyang Technological University,College of Engineering,Aerospace Engineering,95.3,95.3,3494,3500,3536,3500,3100,3816


<p>With reference to the DataFrame output above, we are able to verify that the Duplicated Data Columns have been successfully removed and there are no longer any Duplicated Data Columns visible.</p>

<hr/>
<h3>2.4.2 Identifying Duplicated Data</h3>

<p>Before being able to address Duplicated Data, I have to verify if there are Duplicated Data in the first place. In the cell below, I will be checking for Duplicated Data and based on the output, one of the two action plan will occur.</p>
<br/>
<span style="color:red"><p><strong><u>Duplicated Data Identified</u></strong></p></span>
<ul>
    <li>Proceed to remove the Duplicated Data.</li>
</ul>
<span style="color:green"><p><strong><u>No Duplicated Data Identified</u></strong></p></span>
<ul>
    <li>Proceed with checking next Dataset and / or remove any Duplicated Data found in other Datasets.</li>
</ul>
<br/>
<p>With the action plan listed above, I will be proceeding to check for Duplicated Data.</p>

<hr/>
<h4>2.4.2.1 Identifying Duplicated Data In Dataset 1</h4>

<p>We will proceed to verify the presence of Duplicated Data in the First Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [132]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df1.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,3678


<p>From the output above, I can infer that there is a total of '3678' duplicated data rows. Thus there is <span style="color:red"><strong><u>Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the removal of duplicated data.</p>

<hr/>
<h4>2.4.2.2 Identifying Duplicated Data In Dataset 2</h4>

<p>We will proceed to verify the presence of Duplicated Data in the Second Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [136]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df2.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output above, I can infer that there is a total of '0' duplicated data rows. Thus there is <span style="color:green"><strong><u>No Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the identification of other Datasets.</p>

<hr/>
<h4>2.4.2.3 Identifying Duplicated Data In Dataset 3</h4>

<p>We will proceed to verify the presence of Duplicated Data in the Third Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [140]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df3.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output above, I can infer that there is a total of '0' duplicated data rows. Thus there is <span style="color:green"><strong><u>No Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the identification of other Datasets.</p>

<hr/>
<h4>2.4.2.4 Identifying Duplicated Data In Dataset 4</h4>

<p>We will proceed to verify the presence of Duplicated Data in the Forth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [144]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df4.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output above, I can infer that there is a total of '0' duplicated data rows. Thus there is <span style="color:green"><strong><u>No Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the identification of other Datasets.</p>

<hr/>
<h4>2.4.2.5 Identifying Duplicated Data In Dataset 5</h4>

<p>We will proceed to verify the presence of Duplicated Data in the Fifth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [148]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df5.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output above, I can infer that there is a total of '0' duplicated data rows. Thus there is <span style="color:green"><strong><u>No Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the identification of other Datasets.</p>

<hr/>
<h4>2.4.2.6 Identifying Duplicated Data In Dataset 6</h4>

<p>We will proceed to verify the presence of Duplicated Data in the Sixth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [152]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df6.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output above, I can infer that there is a total of '0' duplicated data rows. Thus there is <span style="color:green"><strong><u>No Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the identification of other Datasets.</p>

<hr/>
<h4>2.4.2.7 Identifying Duplicated Data In Dataset 7</h4>

<p>We will proceed to verify the presence of Duplicated Data in the Seventh Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Duplicated Data / move on to the next procedure.</p>

In [156]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df7.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output above, we can infer that there is a total of '0' duplicated data rows. Thus there is <span style="color:green"><strong><u>No Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed remove the Duplicated Data Rows in the next sub-section.</p>

<hr/>
<h3>2.4.3 Removing Duplicated Data Rows - Dataset 1</h3>

<p>With the presence of Duplicated Data identified and confirmed in Dataset 1, it is essential to remove it as Duplicated Data can distort data analysis causing any analysis done to be inaccurate and unreliable.</p>

In [160]:
# ----------- Removing Duplicated Data ----------- #
df1 = df1.drop_duplicates()

# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = df1.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output of 0 as seen above, I am able to verify that all duplicated data has been successfully removed and will no logner affect any data analysis conducted in the later part.</p>

<hr/>
<h2>2.5 Abnormality Handling</h2>

<p>In order to ensure the Datasets are consistent, I will be identifying and subseqeuntly handle abnormalies within each of the Datasets. Abnormally can come in multiple forms such as placing a placeholder value for Missing Data or Human Error which may have occured due to negilance or simply due to differing practices from Data Entry Clerks.</p>

<hr/>
<h3>2.5.1 Identifying Abnormality</h3>

<p>Before being able to address the Abnormal Data within a Dataset, we will firstly need to identify if there are any Abnormally within the dataset in the first place. The action plan is listed below should any Abnormally be identified.</p>
<br/>
<span style="color:red"><p><strong><u>Abnormality Identified</u></strong></p></span>
<ul>
    <li>Proceed to inspect the Data Row with the Abnormally accordingly and address it depending on what suits best with the Dataset context.</li>
</ul>
<span style="color:green"><p><strong><u>No Abnormality Identified</u></strong></p></span>
<ul>
    <li>Proceed with checking next Dataset and / or address any Abnormality Data found in other Datasets.</li>
</ul>
<br/>
<p>With the action plan listed above, I will be proceeding to check for Abnormality Data.</p>

<hr/>
<h4>2.5.1.1 Identifying Abnormally In Dataset 1 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the First Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [168]:
# ----------- Creating University DataFrame ----------- #
df1_year = pd.DataFrame(np.unique(df1["year"]), columns=["year"])

# ----------- Displaying DataFrame ----------- #
df1_year.style.background_gradient(cmap="Blues")

,year
0,2012
1,2013
2,2014
3,2015
4,2016
5,2017
6,2018
7,2019
8,2020
9,2021


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'university' Data Column as there seems to be a year called 'year'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.2 Identifying Abnormally In Dataset 1 University Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the First Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [172]:
# ----------- Creating University DataFrame ----------- #
df1_universities = pd.DataFrame(np.unique(df1["university"]), columns=["university"])

# ----------- Displaying DataFrame ----------- #
df1_universities.style.background_gradient(cmap="Blues")

,university
0,Nanyang Technological University
1,National University of Singapore
2,Singapore Institute of Technology
3,Singapore Management University
4,Singapore University of Social Sciences
5,Singapore University of Technology and Design
6,university


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'university' Data Column as there seems to be a university called 'university'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.3 Identifying Abnormally In Dataset 1 School Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the First Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [176]:
# ----------- Creating School DataFrame ----------- #
df1_school = pd.DataFrame(np.unique(df1["school"]), columns=["school"])

# ----------- Displaying DataFrame ----------- #
df1_school.style.background_gradient(cmap="Blues")

,school
0,College of Business (Nanyang Business School)
1,College of Design and Engineering
2,College of Engineering
3,"College of Humanities, Arts & Social Sciences"
4,College of Science
5,College of Sciences
6,DigiPen Institute of Technology
7,Faculty Of Dentistry
8,Faculty Of Engineering
9,Faculty of Arts & Social Sciences


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'school' Data Column as there seems to be a school called 'school' and 'na'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.4 Identifying Abnormally In Dataset 2 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Second Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [180]:
# ----------- Convert Numeric to Strings ----------- #
df2['year'] = df2['year'].astype(str)  # Convert all numeric values to strings

# ----------- Apply np.unique ----------- #
unique_years = np.unique(df2['year'])  # Drop NaN before applying np.unique
df2_year = pd.DataFrame(unique_years, columns=["year"])  # Convert unique values to DataFrame

# ----------- Displaying DataFrame ----------- #
df2_year.style.background_gradient(cmap="Blues")

,year
0,1991
1,1992
2,1993
3,1994
4,1996
5,1997
6,1998
7,1999
8,2001
9,2002


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'year' Data Column as there seems to be a school called '2007a' and 'nan' which could just be a missing data. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.5 Identifying Abnormally In Dataset 3 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Third Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [184]:
# ----------- Creating School DataFrame ----------- #
df3_year = pd.DataFrame(np.unique(df3["year"]), columns=["year"])

# ----------- Displaying DataFrame ----------- #
df3_year.style.background_gradient(cmap="Blues")

,year
0,1991
1,1992
2,1993
3,1994
4,1996
5,1997
6,1998
7,1999
8,2001
9,2002


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'year' Data Column as there seems to be a school called '2007a'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.6 Identifying Abnormally In Dataset 3 Sex Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Third Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [188]:
# ----------- Creating School DataFrame ----------- #
df3_sex = pd.DataFrame(np.unique(df3["sex"]), columns=["sex"])

# ----------- Displaying DataFrame ----------- #
df3_sex.style.background_gradient(cmap="Blues")

,sex
0,female
1,male


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'year' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.7 Identifying Abnormally In Dataset 3 Highest Qualification Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Third Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [192]:
# ----------- Creating School DataFrame ----------- #
df3_highest_qualification = pd.DataFrame(np.unique(df3["highest_qualification"]), columns=["highest_qualification"])

# ----------- Displaying DataFrame ----------- #
df3_highest_qualification.style.background_gradient(cmap="Blues")

,highest_qualification
0,degree
1,diploma and professional qualification
2,lower secondary
3,post-secondary (non-tertiary)
4,primary and below
5,secondary


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'highest_qualification' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.8 Identifying Abnormally In Dataset 4 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Forth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [196]:
# ----------- Creating School DataFrame ----------- #
df4_year = pd.DataFrame(np.unique(df4["year"]), columns=["year"])

# ----------- Displaying DataFrame ----------- #
df4_year.style.background_gradient(cmap="Blues")

,year
0,1991
1,1992
2,1993
3,1994
4,1996
5,1997
6,1998
7,1999
8,2001
9,2002


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'year' Data Column as there seems to be a school called '2007a'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.9 Identifying Abnormally In Dataset 4 Sex Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Forth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [200]:
# ----------- Creating School DataFrame ----------- #
df4_sex = pd.DataFrame(np.unique(df4["sex"]), columns=["sex"])

# ----------- Displaying DataFrame ----------- #
df4_sex.style.background_gradient(cmap="Blues")

,sex
0,female
1,male


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'sex' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.10 Identifying Abnormally In Dataset 4 Age Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Forth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [204]:
# ----------- Creating School DataFrame ----------- #
df4_age = pd.DataFrame(np.unique(df4["age"]), columns=["age"])

# ----------- Displaying DataFrame ----------- #
df4_age.style.background_gradient(cmap="Blues")

,age
0,15-19
1,20-24
2,25-29
3,30-34
4,35-39
5,40-44
6,45-49
7,50-54
8,55-59
9,60-64


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'age' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.11 Identifying Abnormally In Dataset 5 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Fifth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [208]:
# ----------- Creating School DataFrame ----------- #
df5_year = pd.DataFrame(np.unique(df5["year"]), columns=["year"])

# ----------- Displaying DataFrame ----------- #
df5_year.style.background_gradient(cmap="Blues")

,year
0,1991
1,1992
2,1993
3,1994
4,1996
5,1997
6,1998
7,1999
8,2001
9,2002


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'year' Data Column as there seems to be a year called '2007a'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.12 Identifying Abnormally In Dataset 5 Sex Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Fifth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [212]:
# ----------- Creating School DataFrame ----------- #
df5_sex = pd.DataFrame(np.unique(df5["sex"]), columns=["sex"])

# ----------- Displaying DataFrame ----------- #
df5_sex.style.background_gradient(cmap="Blues")

,sex
0,female
1,male


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'sex' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.13 Identifying Abnormally In Dataset 5 Age Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Fifth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [216]:
# ----------- Creating School DataFrame ----------- #
df5_age = pd.DataFrame(np.unique(df5["age"]), columns=["age"])

# ----------- Displaying DataFrame ----------- #
df5_age.style.background_gradient(cmap="Blues")

,age
0,15-24
1,25-54
2,25-64
3,55-64
4,65 & Over


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'age' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.14 Identifying Abnormally In Dataset 6 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Sixth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [220]:
# ----------- Creating School DataFrame ----------- #
df6_year = pd.DataFrame(np.unique(df6["Year"]), columns=["Year"])

# ----------- Displaying DataFrame ----------- #
df6_year.style.background_gradient(cmap="Blues")

,Year
0,2019
1,2020
2,2021
3,2022
4,2023


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'Year' Data Column. We will now move on to Identifying Abnormallies for the other Datasets.</p>

<hr/>
<h4>2.5.1.15 Identifying Abnormally In Dataset 6 Gender Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Sixth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [224]:
# ----------- Creating School DataFrame ----------- #
df6_gender = pd.DataFrame(np.unique(df6["Gender"]), columns=["Gender"])

# ----------- Displaying DataFrame ----------- #
df6_gender.style.background_gradient(cmap="Blues")

,Gender
0,Male
1,Female
2,Male


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:red"><strong><u>Abnormality Identified</u></strong></span> in the 'Gender' Data Column as there seems to be 2 gender called 'Male'. We will further investigate this later on.</p>

<hr/>
<h4>2.5.1.16 Identifying Abnormally In Dataset 6 Industry Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Sixth Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [228]:
# ----------- Creating School DataFrame ----------- #
df6_industry = pd.DataFrame(np.unique(df6["Industry"]), columns=["Industry"])

# ----------- Displaying DataFrame ----------- #
df6_industry.style.background_gradient(cmap="Blues")

,Industry
0,"A, B, D, E Others"
1,C Manufacturing
2,F Construction
3,G Wholesale & Retail Trade
4,G-U Services
5,H Transportation & Storage
6,I Accommodation & Food Services
7,J Information & Communications
8,K Financial & Insurance Services
9,L Real Estate Services


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'Industry' Data Column. We will now move on to Identify Abnormallies for the next Dataset.</p>

<hr/>
<h4>2.5.1.17 Identifying Abnormally In Dataset 7 Year Data Column</h4>

<p>We will proceed to verify the presence of Abnormal Data in the Seventh Dataset. Upon the successful execution of the cell below, we will be able to determine which action plan we will need to follow in order to address the Abnormal Data / move on to the next procedure.</p>

In [232]:
# ----------- Creating Year DataFrame ----------- #
df7_year = pd.DataFrame(np.unique(df7["Year"]), columns=["Year"])

# ----------- Displaying DataFrame ----------- #
df7_year.style.background_gradient(cmap="Blues")

,Year
0,1950
1,1951
2,1952
3,1953
4,1954
5,1955
6,1956
7,1957
8,1958
9,1959


<p>With reference to the DataFrame output above, we are able to identify that there is a <span style="color:green"><strong><u>No Abnormality Identified</u></strong></span> in the 'Year' Data Column. We will now move on to Inspection Abnormallies.</p>

<hr/>
<h3>2.5.2 Inspecting Abnormality</h3>

<p>Before being able to address the Abnormal Data within a Dataset, we will firstly need to inspect the abnormalies identified within each of the dataset. This will give us a better idea on whether the abnormally needs to be addressed or not. The inspection also gives us a conclusion on whether the 'abnormally' is indeed an abnormally or just a fluke in the detection     done previously. The action plan for this section has been listed below.</p>
<br/>
<span style="color:red"><p><strong><u>Abnormality Confirmed</u></strong></p></span>
<ul>
    <li>Proceed to address the Abnormally using the most appropriate determined on a case-by-case basis.</li>
</ul>
<span style="color:green"><p><strong><u>False Abnormality</u></strong></p></span>
<ul>
    <li>Proceed with inspect next Dataset and / or address any Abnormality Data found in other Datasets.</li>
</ul>
<br/>
<p>With the action plan listed above, I will be proceeding to inspect the Abnormality Data.</p>

<hr/>
<h4>2.5.2.1 Inspecting Abnormallies In Dataset 1</h4>

<p>As previously identified, Dataset 1 has a number of Abnormalities which are listed below.</p>
<br/>
<ol>
    <li>'year' in the Data Column 'year'</li>
    <li>'university' in the Data Column 'university'</li>
    <li>'school' in the Data Column 'school'</li>
    <li>'na' in the Data Column 'school'</li>
</ol>
<br/>
<p>With the Abnormallies listed out, we will proceed to inspect the Dataset's Abnormalities.</p>

In [238]:
# ----------- Displaying DataFrame With Filters ----------- #
df1[df1["university"] == "university"].style.background_gradient(cmap="Blues")

,year,university,school,degree,employment_rate_overall,employment_rate_ft_perm,basic_monthly_mean,basic_monthly_median,gross_monthly_mean,gross_monthly_median,gross_mthly_25_percentile,gross_mthly_75_percentile
1262,year,university,school,degree,employment_rate_overall,employment_rate_ft_perm,basic_monthly_mean,basic_monthly_median,gross_monthly_mean,gross_monthly_median,gross_mthly_25_percentile,gross_mthly_75_percentile


<p>With reference to the output above, we are able to identify the abnormally 1, 2 and 3. This abnormally most likely happened due to Human Error during concatenation in an external software (e.g. Excel). Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Drop The Data Row</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

In [240]:
# ----------- Creating School Abnormally DataFrame ----------- #
df1_school_abnormally = pd.DataFrame(df1[df1["school"] == "na"])

# ----------- Displaying DataFrame ----------- #
df1_school_abnormally.style.background_gradient(cmap="Blues")

,year,university,school,degree,employment_rate_overall,employment_rate_ft_perm,basic_monthly_mean,basic_monthly_median,gross_monthly_mean,gross_monthly_median,gross_mthly_25_percentile,gross_mthly_75_percentile
285,2015,Singapore University of Technology and Design,na,Bachelor of Engineering (Engineering Product Development),92.5,88.1,3595,3618,3565,3600,3400,3800
286,2015,Singapore University of Technology and Design,na,Bachelor of Engineering (Engineering Systems and Design),80.3,71.8,3699,3450,3763,3500,3190,4100
287,2015,Singapore University of Technology and Design,na,Bachelor of Engineering (Information Systems Technology and Design),86,83.7,3872,3700,3952,3950,3338,4450
288,2015,Singapore University of Technology and Design,na,Bachelor of Science (Architecture and Sustainable Design) ###,na,na,na,na,na,na,na,na
434,2016,Singapore University of Technology and Design,na,Bachelor of Engineering (Engineering Product Development),87,79.7,3620,3500,3690,3600,3300,3750
435,2016,Singapore University of Technology and Design,na,Bachelor of Engineering (Information Systems Technology and Design),91.9,91.9,4197,4000,4232,4000,3500,4950
436,2016,Singapore University of Technology and Design,na,Bachelor of Engineering (Engineering Systems and Design),100,93.5,3695,3600,3830,3650,3550,4200
437,2016,Singapore University of Technology and Design,na,Bachelor of Science (Architecture and Sustainable Design) ###,na,na,na,na,na,na,na,na
566,2017,Singapore University of Technology and Design,na,Bachelor of Engineering (Engineering Product Development),89.9,87.3,3594,3600,3638,3600,3450,3800
567,2017,Singapore University of Technology and Design,na,Bachelor of Engineering (Information Systems Technology and Design),100,100,4146,4000,4175,4023,3600,4900


<p>With reference to the output above, we are able to verify that the school that are filled with the placeholder value of 'na' is found in the 'school' Data Column of Singapore University of Technology and Design. However, we are also able to observe that other Data Column also contain the 'na' placeholder. This could be due to practices done by the Data Entry Clerks so as to fill up the blanks in a 3rd Party Software (e.g. Excel). Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Fill The School With Reference To Degree</u></strong> and <strong><u>Replace 'na' Values On Numerical Data With Numpy NaN</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

In [242]:
# ----------- Creating School DataFrame ----------- #
df1_school = pd.DataFrame(np.unique(df1[df1["school"] == "na"]["degree"]), columns=["degree"])

# ----------- Displaying DataFrame ----------- #
df1_school.style.background_gradient(cmap="Blues")

,degree
0,Bachelor of Engineering (Computer Science and Design)
1,Bachelor of Engineering (Engineering Product Development)
2,Bachelor of Engineering (Engineering Systems and Design)
3,Bachelor of Engineering (Information Systems Technology and Design)
4,Bachelor of Science (Architecture and Sustainable Design)
5,Bachelor of Science (Architecture and Sustainable Design) ###


<p>With reference to the output of the DataFrame above, I am able to map out the placeholder value to put within the 'school' Data Column. The mapping is done below.</p>
<br/>
<p><strong>Bachelor of Engineering (Computer Science and Design)</strong> &#10132; <i>School of Computing</i></p>
<p><strong>Bachelor of Engineering (Engineering Product Development)</strong> &#10132; <i>School of Computing</i></p>
<p><strong>Bachelor of Engineering (Engineering Systems and Design)</strong> &#10132; <i>School of Engineering</i></p>
<p><strong>Bachelor of Engineering (Information Systems Technology and Design)</strong> &#10132; <i>School of Computing</i></p>
<p><strong>Bachelor of Science (Architecture and Sustainable Design)</strong> &#10132; <i>School of Science</i></p>
<p><strong>Bachelor of Science (Architecture and Sustainable Design) ###</strong> &#10132; <i>School of Science</i></p>
<br/>
<p>With the mapping planned out above, we will proceed to fill the Missing Data.</p>

<hr/>
<h4>2.5.2.2 Inspecting Abnormallies In Dataset 2</h4>

<p>As previously identified, Dataset 2 has a number of Abnormalities which are listed below.</p>
<br/>
<ol>
    <li>'2007a' in the Data Column 'year'</li>
    <li>'nan' value in the Data Column 'year</li>
</ol>
<br/>
<p>With the Abnormallies listed out, we will proceed to inspect the Dataset's Abnormalities.</p>

In [246]:
# ----------- Displaying DataFrame With Filters ----------- #
df2[df2["year"] == "2007a"].style.background_gradient(cmap="Blues")

,year,overall_labour_force,resident_labour_force,Unnamed: 3,Unnamed: 4,Unnamed: 5
14,2007a,2710300.000000,1878000.000000,nan,nan,nan


<p>With reference to the output of the DataFrame above, we are able to see that the Data Column with the year '2007a' does have Valid Data. However, as the year is still recognised as invalid, we will still have to address the Data Row as 2007 already exist. From the output of the DataFrame, we are also able to see the 3 invalid Data Columns which are 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5. These Data Columns contain no values at all and are most likely due to Human Error where additional Data Columns were created by accident. Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Drop The Data Row Containing '2007a'</u></strong> and <strong><u>Drop All Invalid Data Columns</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

In [248]:
# ----------- Filtering Rows With NaN Data ----------- #
nan_rows = df2[df2.isna().any(axis=1)]

# ----------- Displaying DataFrame With Filters ----------- #
nan_rows.style.background_gradient(cmap="Blues")

,year,overall_labour_force,resident_labour_force,Unnamed: 3,Unnamed: 4,Unnamed: 5
0,1991,1673700.000000,1372900.000000,nan,nan,nan
1,1992,1733600.000000,1409900.000000,nan,nan,nan
2,1993,1762700.000000,1421700.000000,nan,nan,nan
3,1994,1842200.000000,1456100.000000,nan,nan,nan
4,1996,2024900.000000,1511500.000000,nan,nan,nan
5,1997,2116000.000000,1538300.000000,nan,nan,nan
6,1998,2187900.000000,1546500.000000,nan,nan,nan
7,1999,2208700.000000,1595900.000000,nan,nan,nan
8,2001,2330500.000000,1644300.000000,nan,nan,nan
9,2002,2320600.000000,1667900.000000,nan,nan,nan


<p>With reference to the output above, we are able to verify that the last Data Row of the Dataset is all 'NaN'. This is most likely due to a human error where the Data Entry Clerk accidentally added an additional row without noticing. We are also able to verify that the Invalid Data Columns are all filled with 'NaN' and noi legitimate data. Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Drop The Last Data Row Containing 'NaN'</u></strong> and <strong><u>Drop All Invalid Data Columns</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

<hr/>
<h4>2.5.2.3 Inspecting Abnormallies In Dataset 3</h4>

<p>As previously identified, Dataset 3 has a number of Abnormalities which are listed below.</p>
<br/>
<ol>
    <li>'2007a' in the Data Column 'year'</li>
</ol>
<br/>
<p>With the Abnormallies listed out, we will proceed to inspect the Dataset's Abnormalities.</p>

In [252]:
# ----------- Displaying DataFrame With Filters ----------- #
df3[df3["year"] == "2007a"].style.background_gradient(cmap="Blues")

,year,sex,highest_qualification,labour_force
168,2007a,male,primary and below,172900
169,2007a,male,lower secondary,157600
170,2007a,male,secondary,237000
171,2007a,male,post-secondary (non-tertiary),99000
172,2007a,male,diploma and professional qualification,164600
173,2007a,male,degree,247400
174,2007a,female,primary and below,111800
175,2007a,female,lower secondary,81700
176,2007a,female,secondary,213900
177,2007a,female,post-secondary (non-tertiary),66100


<p>With reference to the output of the DataFrame above, we are able to see that the Data Column with the year '2007a' does have Valid Data. However, as the year is still recognised as invalid, we will still have to address the Data Row as 2007 already exist. Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Drop The Data Row Containing '2007a'</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

<hr/>
<h4>2.5.2.4 Inspecting Abnormallies In Dataset 4</h4>

<p>As previously identified, Dataset 4 has a number of Abnormalities which are listed below.</p>
<br/>
<ol>
    <li>'2007a' in the Data Column 'year'</li>
</ol>
<br/>
<p>With the Abnormallies listed out, we will proceed to inspect the Dataset's Abnormalities.</p>

In [256]:
# ----------- Displaying DataFrame With Filters ----------- #
df4[df4["year"] == "2007a"].style.background_gradient(cmap="Blues")

,year,sex,age,resident_labour_force
337,2007a,male,20-24,73300
338,2007a,male,25-29,104000
339,2007a,male,30-34,129900
340,2007a,male,35-39,141700
341,2007a,male,40-44,148600
342,2007a,male,45-49,151500
343,2007a,male,50-54,132400
344,2007a,male,55-59,97400
345,2007a,male,60-64,45100
346,2007a,male,65-69,21600


<p>With reference to the output of the DataFrame above, we are able to see that the Data Column with the year '2007a' does have Valid Data. However, as the year is still recognised as invalid, we will still have to address the Data Row as 2007 already exist. Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Drop The Data Row Containing '2007a'</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

<hr/>
<h4>2.5.2.5 Inspecting Abnormallies In Dataset 5</h4>

<p>As previously identified, Dataset 5 has a number of Abnormalities which are listed below.</p>
<br/>
<ol>
    <li>'2007a' in the Data Column 'year'.</li>
</ol>
<br/>
<p>With the Abnormallies listed out, we will proceed to inspect the Dataset's Abnormalities.</p>

In [260]:
# ----------- Displaying DataFrame With Filters ----------- #
df5[df5["year"] == "2007a"].style.background_gradient(cmap="Blues")

,year,sex,age,resident_labour_force_participation_rate
140,2007a,male,25-64,93.000000
141,2007a,male,15-24,37.900000
142,2007a,male,25-54,96.400000
143,2007a,male,55-64,77.400000
144,2007a,male,65 & Over,23.500000
145,2007a,female,25-64,66.000000
146,2007a,female,15-24,35.700000
147,2007a,female,25-54,71.600000
148,2007a,female,55-64,39.900000
149,2007a,female,65 & Over,8.000000


<p>With reference to the output of the DataFrame above, we are able to see that the Data Column with the year '2007a' does have Valid Data. However, as the year is still recognised as invalid, we will still have to address the Data Row as 2007 already exist. Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Drop The Data Row Containing '2007a'</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

<hr/>
<h4>2.5.2.6 Inspecting Abnormallies In Dataset 6</h4>

<p>As previously identified, Dataset 6 has a number of Abnormalities which are listed below.</p>
<br/>
<ol>
    <li>Duplicated 'Male' in the Data Column 'Gender'.</li>
</ol>
<br/>
<p>With the Abnormallies listed out, we will proceed to inspect the Dataset's Abnormalities.</p>

In [264]:
# ----------- Displaying DataFrame With Filters ----------- #
df6[df6["Gender"] == " Male"].style.background_gradient(cmap="Blues")

,Amt,Gender,Industry,Year
96,5322,Male,C Manufacturing,2020
97,4563,Male,F Construction,2020
98,4680,Male,G-U Services,2020
99,4652,Male,G Wholesale & Retail Trade,2020
100,2543,Male,H Transportation & Storage,2020
101,2535,Male,I Accommodation & Food Services,2020
102,7020,Male,J Information & Communications,2020
103,8020,Male,K Financial & Insurance Services,2020
104,4056,Male,L Real Estate Services,2020
105,6581,Male,M Professional Services,2020


<p>With reference to the output of the DataFrame above, we are able to see that the Data Column with the year ' Male' does have Valid Data. Therefore, the result of the inspection is <span style="color:red"><strong><u>Abnormality Confirmed</u></strong></span> and the best course of action determined here is to <strong><u>Correct The Data Row Containing ' Male'</u></strong>. This operation will be done in the next section after all the inspection is done.</p>

<hr/>
<h3>2.5.3 Addressing Abnormality</h3>

<p>In this section, we will be addressing the abnormallies identified and inspected in the previous sections. The modus operandi that we will utilise will be the same as what we deemed suitable in the previous section. The summarised actions that will be taken are listed below.</p>
<br/>
<p><strong><u>Dataset 1</u></strong></p>
<ul>
    <li>Map Out School With Reference To Degree.</li>
    <li>Replace 'na' With Numpy NaN.</li>
    <li>Remove Error Row.</li>
</ul>
<p><strong><u>Dataset 2</u></strong></p>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
    <li>Remove All Invalid Data Columns.</li>
    <li>Remove Last Data Row.</li>
</ul>
<p><strong><u>Dataset 3</u></strong></p>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
</ul>
<p><strong><u>Dataset 4</u></strong></p>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
</ul>
<p><strong><u>Dataset 5</u></strong></p>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
</ul>
<p><strong><u>Dataset 6</u></strong></p>
<ul>
    <li>Correct Data Row With The Gender ' Male'.</li>
</ul>
<br/>
<p>With each action plan for the respective Datasets listed out, we will proceed to address the adnormally accordingly.</p>

<hr/>
<h4>2.5.3.1 Addressing Abnormallies In Dataset 1</h4>

<p>To address the Abnormallies in Dataset 1, we will follow the list of operations that needs to be done. This list is indicated below and we will follow it as a strict guideline.</p>
<br/>
<ul>
    <li>Map Out School With Reference To Degree.</li>
    <li>Replace 'na' With Numpy NaN.</li>
    <li>Remove Error Row.</li>
</ul>
<br/>
<p>With the list of things to be accomplished listed out, we will proceed to perform the necessary operation in the cell below.</p>

In [270]:
# ----------- Mapping Dictionary For School Data Column ----------- #
degree_to_school = {
    "Bachelor of Engineering (Computer Science and Design)": "School of Computing",
    "Bachelor of Engineering (Engineering Product Development)": "School of Computing",
    "Bachelor of Engineering (Engineering Systems and Design)": "School of Engineering",
    "Bachelor of Engineering (Information Systems Technology and Design)": "School of Computing",
    "Bachelor of Science (Architecture and Sustainable Design)": "School of Science ",
    "Bachelor of Science (Architecture and Sustainable Design) ###": "School of Science"
}

# ----------- Applying Mapping To School Data Column ----------- #
df1['school'] = df1['school'].replace('na', np.nan)
df1['school'] = df1['school'].fillna(df1['degree'].map(degree_to_school))

# ----------- Replacing 'na' with Numpy NaN ----------- #
df1['employment_rate_overall'] = df1['employment_rate_overall'].replace('na', np.nan)
df1['employment_rate_ft_perm'] = df1['employment_rate_ft_perm'].replace('na', np.nan)
df1['basic_monthly_mean'] = df1['basic_monthly_mean'].replace('na', np.nan)
df1['basic_monthly_median'] = df1['basic_monthly_median'].replace('na', np.nan)
df1['gross_monthly_mean'] = df1['gross_monthly_mean'].replace('na', np.nan)
df1['gross_monthly_median'] = df1['gross_monthly_median'].replace('na', np.nan)
df1['gross_mthly_25_percentile'] = df1['gross_mthly_25_percentile'].replace('na', np.nan)
df1['gross_mthly_75_percentile'] = df1['gross_mthly_75_percentile'].replace('na', np.nan)

# ----------- Removing Error Row ----------- #
df1 = df1[df1["university"] != "university"]

# ----------- Displaying DataFrame ----------- #
df1.head(5).style.background_gradient(cmap="Blues")

,year,university,school,degree,employment_rate_overall,employment_rate_ft_perm,basic_monthly_mean,basic_monthly_median,gross_monthly_mean,gross_monthly_median,gross_mthly_25_percentile,gross_mthly_75_percentile
0,2013,Nanyang Technological University,College of Business (Nanyang Business School),Accountancy and Business,97.4,96.1,3701,3200,3727,3350,2900,4000
1,2013,Nanyang Technological University,College of Business (Nanyang Business School),Accountancy (3-yr direct Honours Programme),97.1,95.7,2850,2700,2938,2700,2700,2900
2,2013,Nanyang Technological University,College of Business (Nanyang Business School),Business (3-yr direct Honours Programme),90.9,85.7,3053,3000,3214,3000,2700,3500
3,2013,Nanyang Technological University,College of Business (Nanyang Business School),Business and Computing,87.5,87.5,3557,3400,3615,3400,3000,4100
4,2013,Nanyang Technological University,College of Engineering,Aerospace Engineering,95.3,95.3,3494,3500,3536,3500,3100,3816


<p>With reference to the output of the DataFrame above, we are able to verify that the Abnormallies have been successfully addressed and we are able to move on to the next Dataset that needs to have its Abnormallies addressed.</p>

<hr/>
<h4>2.5.3.2 Addressing Abnormallies In Dataset 2</h4>

<p>To address the Abnormallies in Dataset 2, we will follow the list of operations that needs to be done. This list is indicated below and we will follow it as a strict guideline.</p>
<br/>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
    <li>Remove All Invalid Data Columns.</li>
    <li>Remove Last Data Row.</li>
</ul>
<br/>
<p>With the list of things to be accomplished listed out, we will proceed to perform the necessary operation in the cell below.</p>

In [274]:
# ----------- Removing Error Row ----------- #
df2 = df2[df2["year"] != "2007a"]

# ---------- Dropping Duplicated Data Columns ---------- #
df2.drop(["Unnamed: 3", "Unnamed: 4", "Unnamed: 5"], axis=1, inplace=True)

# ----------- Removing Last Row Of DataFrame ----------- #
df2 = df2.iloc[:-1]

# ----------- Displaying DataFrame ----------- #
df2.head(5).style.background_gradient(cmap="Blues")

,year,overall_labour_force,resident_labour_force
0,1991,1673700.000000,1372900.000000
1,1992,1733600.000000,1409900.000000
2,1993,1762700.000000,1421700.000000
3,1994,1842200.000000,1456100.000000
4,1996,2024900.000000,1511500.000000


<p>With reference to the output of the DataFrame above, we are able to verify that the Abnormallies have been successfully addressed and we are able to move on to the next Dataset that needs to have its Abnormallies addressed.</p>

<hr/>
<h4>2.5.3.3 Addressing Abnormallies In Dataset 3</h4>

<p>To address the Abnormallies in Dataset 3, we will follow the list of operations that needs to be done. This list is indicated below and we will follow it as a strict guideline.</p>
<br/>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
</ul>
<br/>
<p>With the list of things to be accomplished listed out, we will proceed to perform the necessary operation in the cell below.</p>

In [278]:
# ----------- Removing Error Row ----------- #
df3 = df3[df3["year"] != "2007a"]

# ----------- Displaying DataFrame ----------- #
df3.head(5).style.background_gradient(cmap="Blues")

,year,sex,highest_qualification,labour_force
0,1991,male,primary and below,281300
1,1991,male,lower secondary,158800
2,1991,male,secondary,224400
3,1991,male,post-secondary (non-tertiary),76000
4,1991,male,diploma and professional qualification,58300


<p>With reference to the output of the DataFrame above, we are able to verify that the Abnormallies have been successfully addressed and we are able to move on to the next Dataset that needs to have its Abnormallies addressed.</p>

<hr/>
<h4>2.5.3.4 Addressing Abnormallies In Dataset 4</h4>

<p>To address the Abnormallies in Dataset 4, we will follow the list of operations that needs to be done. This list is indicated below and we will follow it as a strict guideline.</p>
<br/>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
</ul>
<br/>
<p>With the list of things to be accomplished listed out, we will proceed to perform the necessary operation in the cell below.</p>

In [282]:
# ----------- Removing Error Row ----------- #
df4 = df4[df4["year"] != "2007a"]

# ----------- Displaying DataFrame ----------- #
df4.head(5).style.background_gradient(cmap="Blues")

,year,sex,age,resident_labour_force
0,1991,male,15-19,36900
1,1991,male,20-24,99400
2,1991,male,25-29,134500
3,1991,male,30-34,143300
4,1991,male,35-39,131000


<p>With reference to the output of the DataFrame above, we are able to verify that the Abnormallies have been successfully addressed and we are able to move on to the next Dataset that needs to have its Abnormallies addressed.</p>

<hr/>
<h4>2.5.3.5 Addressing Abnormallies In Dataset 5</h4>

<p>To address the Abnormallies in Dataset 5, we will follow the list of operations that needs to be done. This list is indicated below and we will follow it as a strict guideline.</p>
<br/>
<ul>
    <li>Remove Data Row With The Year '2007a'.</li>
</ul>
<br/>
<p>With the list of things to be accomplished listed out, we will proceed to perform the necessary operation in the cell below.</p>

In [286]:
# ----------- Removing Error Row ----------- #
df5 = df5[df5["year"] != "2007a"]

# ----------- Displaying DataFrame ----------- #
df5.head(5).style.background_gradient(cmap="Blues")

,year,sex,age,resident_labour_force_participation_rate
0,1991,male,25-64,92.600000
1,1991,male,15-24,55.900000
2,1991,male,25-54,97.500000
3,1991,male,55-64,59.700000
4,1991,male,65 & Over,19.900000


<p>With reference to the output of the DataFrame above, we are able to verify that the Abnormallies have been successfully addressed and we are able to move on to the next Dataset that needs to have its Abnormallies addressed.</p>

<hr/>
<h4>2.5.3.6 Addressing Abnormallies In Dataset 6</h4>

<p>To address the Abnormallies in Dataset 6, we will follow the list of operations that needs to be done. This list is indicated below and we will follow it as a strict guideline.</p>
<br/>
<ul>
    <li>Correct Data Row With The Gender ' Male'.</li>
</ul>
<br/>
<p>With the list of things to be accomplished listed out, we will proceed to perform the necessary operation in the cell below.</p>

In [290]:
# ----------- Correcting Gender ----------- #
df6["Gender"] = df6["Gender"].str.strip().str.lower().str.capitalize()

# ----------- Defining DataFrame For Gender ----------- #
df6_gender = pd.DataFrame(df6["Gender"].unique(), columns=["Gender"])

# ----------- Displaying DataFrame ----------- #
df6_gender.style.background_gradient(cmap="Blues")

,Gender
0,Male
1,Female


<p>With reference to the output of the DataFrame above, we are able to verify that the Abnormallies have been successfully addressed and we are able to move on to the next Dataset that needs to have its Abnormallies addressed.</p>

<hr/>
<h2>2.6 Fixing Data Types</h2>